pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return nil, err
+	}
+	// Find the besteffort pods we would need to evict if we already evicted the required guaranteed and burstable pods.
+	bestEffortToEvict, err := getPodsToPreemptByDistance(bestEffortPods, requirements.subtract(append(burstableToEvict, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	return append(append(bestEffortToEvict, burstableToEvict...), guarateedToEvict...), nil
+}
+
+// finds the pods that have pod requests >= admission requirements.
+// Chooses pods that minimize "distance" to the requirements.
+// If more than one pod exists that fulfills the remaining requirements,
+// it chooses the pod that has the "smaller resource request"
+// This method, by repeatedly choosing the pod that fulfills as much of the requirements as possible,
+// attempts to minimize the number of pods returned.
+func getPodsToPreemptByDistance(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	podsToEvict := []*v1.Pod{}
+	// evict pods by shortest distance from remaining requirements, updating requirements every round.
+	for len(requirements) > 0 {
+		if len(pods) == 0 {
+			return nil, fmt.Errorf("no set of running pods found to reclaim resources: %v", requirements.toString())
+		}
+		// all distances must be less than len(requirements), because the max distance for a single requirement is 1
+		bestDistance := float64(len(requirements) + 1)
+		bestPodIndex := 0
+		// Find the pod with the smallest distance from requirements
+		// Or, in the case of two equidistant pods, find the pod with "smaller" resource requests.
+		for i, pod := range pods {
+			dist := requirements.distance(pod)
+			if dist < bestDistance || (bestDistance == dist && smallerResourceRequest(pod, pods[bestPodIndex])) {
+				bestDistance = dist
+				bestPodIndex = i
+			}
+		}
+		// subtract the pod from requirements, and transfer the pod from input-pods to pods-to-evicted
+		requirements = requirements.subtract(pods[bestPodIndex])
+		podsToEvict = append(podsToEvict, pods[bestPodIndex])
+		pods[bestPodIndex] = pods[len(pods)-1]
+		pods = pods[:len(pods)-1]
+	}
+	return podsToEvict, nil
+}
+
+type admissionRequirement struct {
+	resourceName v1.ResourceName
+	quantity     int64
+}
+
+type admissionRequirementList []*admissionRequirement
+
+// distance of the pods requests from the admissionRequirements.
+// distance is measured by the fraction of the requirement satisfied by the pod,
+// so that each requirement is weighted equally, regardless of absolute magnitude.
+func (a admissionRequirementList) distance(pod *v1.Pod) float64 {
+	dist := float64(0)
+	for _, req := range a {
+		remainingRequest := float64(req.quantity - v1.GetResourceRequest(pod, req.resourceName))
+		if remainingRequest < 0 {
+			remainingRequest = 0
+		}
+		dist += math.Pow(remainingRequest/float64(req.quantity), 2)
```

Comment:
(2) is hard to know what's better without real data, and this will vary depending on people's workloads. I've heard some talk of having auto-tuning for resource requests (based on Pod profiling), if we had that then I think preemption behavior would be an interesting to auto-tune too. But I'll say that your intuition feels right to me as well.

(1) Yeah I guess you're right. Since you add over the number of requirements normalizing ends up just multiplying a constant factor over the inequality when you compare two pods, and you get the same ordering either way.